lib/Expr/ArrayExprOptimizer.cpp

```diff
@@ -163,9 +163,10 @@ bool ExprOptimizer::computeIndexes(array2idx_ty &arrays, const ref<Expr> &e,
       // skipping all those indexes that are not multiple of such value.
       // In fact, they will be rejected by the MulExpr interpreter since it
       // will not find any integer solution
-      Expr &e = *idxt_v.getMul();
-      auto &ce = static_cast<ConstantExpr &>(e);
-      uint64_t mulVal = (*ce.getAPValue().getRawData());
+      auto e = idxt_v.getMul();
+      auto ce = dyn_cast<ConstantExpr>(e);
+      assert(ce && "Not a constant expression");
+      uint64_t mulVal = (*ce->getAPValue().getRawData());
```

Comment:
I assume the assumption here is that the expression is constant.  But I guess a dynamic cast is not a performance issue in this context.

Reply:
It should not be a performance issue: it's `dyn_cast<>` (from LLVM: http://llvm.org/docs/HowToSetUpLLVMStyleRTTI.html) not `dynamic_cast` (that needs a vtable).

The problem with the original was that it always assumes a `ConstantExpr` which might be not always true. But we would not detect this with `static_cast`.

Reply:
Thanks, I did read this as a `dynamic_cast`